keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1307 @@
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+
+### Risks and Mitigations
+
+Risk: Our encoder produces some output which is not valid YAML.
+Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.
+
+Risk: The result could be net worse than plain YAML.
+Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.
+
+Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
+Mitigation: Docs?
+
+Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
+Mitigation: Be as precise as possible where we are confident, but leave room where we are not.
+
+## Design Details
+
+Creating an encoder for this format is not very technically challenging,
+but there are many details to consider.
+
+### Packaging
+
+<<[UNRESOLVED where it lives]>>
+
+The simplest thing to do is to write a new implementation of
+k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in
+staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most
+obviously reusable place for it.
+
+A more ecosystem-friendly solution is to write a new function or package under
+sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.
+
+<<[/UNRESOLVED]>>
+
+### Tooling
+
+k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
+ensures they are formatted “correctly”. The only meaning of “correct” is “block
+style”, and it corrupts KYAML input today (aside: it also does not handle
+multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML
+output, and set the default to KYAML (or to change the parent script to always
+specify KYAML).
+
+This tooling will preserve comments.
+
+#### How?
+
+Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
+KYAML is an output format, our tooling goals require that we get access to a
+YAML AST, which may require some dirty tricks.  If we ever change YAML
+libraries (e.g.
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may
+need different dirty tricks.
+
+Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST,
+which we can render as KYAML.  Neither, however, offers a way to encode an
+arbitrary Go object as the equivalent YAML AST.  To do that we may have to
+encode the object to plain YAML, then decode it to AST, then render it as
+KYAML.  That extra conversion is ugly, but this is not generally
+performance-sensitive code and is not run on the server.
+
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an
```

Comment:
what is the impact of this dependency? is this going to impact k8s.io/{api, apimachinery or client-go}

Reply:
It's a significant decision

Reply:
If we predicate this KEP on adopting goccy I am afraid it will take longer than a single release :)

The amount of fear and defense-in-depth we have established around go-yaml is significant.  @liggitt